###<font color="red">Latihan Soal
####<font color="red">A. Pilihan Ganda
<font color="black">  

1. Struktur data yang menjadikan pencarian bersifat melebar level demi level adalah… (a)
tumpukan; (b) antrian; (c) pohon biner; (d) dictionary.
2. Jaminan yang diberikan BFS pada graf tanpa bobot adalah menemukan… (a) rute ber-km
terpendek; (b) rute dengan jumlah langkah tersedikit; (c) semua rute; (d) rute yang sama dengan
DFS.
3. Hasil DFS dapat berubah bila… (a) heuristiknya diganti; (b) urutan tetangga dalam penyimpanan
graf diubah; (c) bobot sisi diubah; (d) tujuan lebih dekat.
4. Heuristik disebut admissible bila… (a) selalu nol; (b) selalu lebih besar dari biaya sebenarnya;
(c) tidak pernah melebih-lebihkan biaya sebenarnya; (d) sama untuk semua simpul.
5. Mencari rute pengiriman dengan bobot waktu tempuh dan tersedia perkiraan garis lurus —
metode paling tepat: (a) BFS; (b) DFS; (c) A*; (d) pencarian acak.

**JAWABAN**
1. b. antrian  
2. b. rute dengan jumlah langkah tersedikit  
3. b. urutan tetangga dalam penyimpanan graf diubah  
4. c. tidak pernah melebih-lebihkan biaya sebenarnya  
5. c.  A*
-----------------------------------------------------------------------------
####<font color="red">B. Esai
<font color="black">

- 1. Buat trace table BFS untuk graf: A–B, A–C, B–D, C–D, D–E (mulai A, tujuan E; urutan tetangga sesuai abjad). Tuliskan jalur akhirnya dan jumlah langkahnya.
- 2. Dengan tabel jarak dan h pada Subbab 2.5, kerjakan dua langkah pertama A* untuk rute  
Ponorogo → Surabaya: tuliskan isi daftar terbuka beserta nilai f-nya di tiap langkah.
- 3. BFS menemukan rute 248 km, A* menemukan 156 km — padahal keduanya sama-sama
“benar”. Jelaskan dalam bahasa sendiri apa yang sebenarnya dioptimalkan masing-masing, dan
kapan jawaban BFS justru yang dibutuhkan.
- 4. Mengapa BFS boros memori sedangkan DFS hemat? Kaitkan dengan apa yang harus disimpan
masing-masing di daftar tunggunya.
- 5. Seorang praktikan mengganti h Surabaya-nya menjadi 100 (padahal jarak sebenarnya 0 karena
Surabaya adalah tujuan). Analisis: apakah h ini masih admissible? Apa akibat yang mungkin
terjadi pada hasil A*?

**JAWABAN**

1. BFS dimulai dari A, lalu B dan C masuk antrian. B diproses duluan, D ditemukan lewat B dan masuk antrian. C diproses, tapi tetangganya (A, D) sudah pernah dilihat jadi tidak ada tambahan. D diproses, E ditemukan dan masuk antrian. Terakhir E diproses dan ternyata itu tujuannya, pencarian selesai. Jalur yang ditemukan: A → B → D → E, sebanyak 3 langkah.

2. Langkah pertama Ponorogo diproses dengan g=0, h=160, f=160, lalu Madiun masuk daftar terbuka dengan g=29, h=140, f=169. Langkah kedua, Madiun diproses, menghasilkan Nganjuk (g=76, h=100, f=176) dan Ngawi (g=61, h=155, f=216) yang masuk daftar terbuka. Simpul dengan f terkecil untuk dikembangkan berikutnya adalah Nganjuk.

3. BFS mengoptimalkan jumlah langkah, sedangkan A* mengoptimalkan total bobot atau jarak nyata karena memakai g (biaya sesungguhnya) dan h (perkiraan sisa). Keduanya sama-sama benar sesuai tujuan masing-masing. Jawaban BFS lebih dibutuhkan saat setiap langkah punya bobot yang sama pentingnya, misalnya menghitung jumlah lompatan pada puzzle geser angka atau jumlah klik pada navigasi menu.

4. BFS menyimpan seluruh simpul di satu level penuh dalam antriannya sebelum lanjut ke level berikutnya, sedangkan DFS hanya menyimpan satu jalur aktif di tumpukannya sampai mentok. Menyimpan satu level penuh jauh lebih besar dibanding menyimpan satu lintasan saja, sehingga BFS lebih boros memori.

5. Heuristik tersebut tidak admissible karena biaya sebenarnya dari tujuan ke tujuan pasti 0, bukan 100, sehingga melebih-lebihkan. Akibatnya f(Surabaya) tampak lebih mahal dari yang sebenarnya, membuat A* menunda memproses Surabaya dan berpotensi kehilangan jaminan menemukan rute paling optimal.

### <font color="red">Kegiatan 1</font>

In [ ]:
# Tanpa deklarasi tipe -- Python menebak sendiri dari isinya
nama = "Budi"       # str  (string)
umur = 19           # int  (bilangan bulat)
ipk = 3.75          # float (pecahan)
aktif = True        # bool (True/False)
print(type(nama), type(umur), type(ipk), type(aktif))

# printf("%s berumur %d tahun, IPK %.2f", nama, umur, ipk);
print(f"{nama} berumur {umur} tahun, IPK {ipk:.2f}")

<class 'str'> <class 'int'> <class 'float'> <class 'bool'>
Budi berumur 19 tahun, IPK 3.75


### <font color="red">Adaptasi kegiatan 1</font>

In [ ]:
# Adaptasi: variabel dan f-string untuk data nilai mahasiswa
nama_mhs = "Farid"      # str
nilai_uts = 85          # int
rata_rata = 78.5        # float
lulus = True            # bool
print(type(nama_mhs), type(nilai_uts), type(rata_rata), type(lulus))

print(f"{nama_mhs} mendapat nilai UTS {nilai_uts}, rata-rata kelas {rata_rata:.2f}")

<class 'str'> <class 'int'> <class 'float'> <class 'bool'>
Farid mendapat nilai UTS 85, rata-rata kelas 78.50


### <font color="red">Kegiatan 2</font>

### <font color="red">C++ sebagai Pembanding dengan Python</font>
```cpp
for (int i = 1; i <= 10; i++) {
  if (i % 2 == 0) {
    printf("%d genap\n", i);
  } else {
    printf("%d ganjil\n", i);
  }
}
```

### <font color="red">Python Code</font>


In [ ]:
for i in range(1, 11):
    if i % 2 == 0:
        print(f"{i} genap")
    else:
        print(f"{i} ganjil")

1 ganjil
2 genap
3 ganjil
4 genap
5 ganjil
6 genap
7 ganjil
8 genap
9 ganjil
10 genap


### <font color="red">Adaptasi kegiatan 2</font>

In [ ]:
# Adaptasi: perulangan dan percabangan untuk mengecek status kelulusan nilai 1-10 mahasiswa
nilai_list = [85, 40, 67, 92, 55, 78, 30, 88, 60, 45]

for i, nilai in enumerate(nilai_list, start=1):
    if nilai >= 60:
        print(f"Mahasiswa ke-{i}: nilai {nilai} -> LULUS")
    else:
        print(f"Mahasiswa ke-{i}: nilai {nilai} -> REMIDI")

Mahasiswa ke-1: nilai 85 -> LULUS
Mahasiswa ke-2: nilai 40 -> REMIDI
Mahasiswa ke-3: nilai 67 -> LULUS
Mahasiswa ke-4: nilai 92 -> LULUS
Mahasiswa ke-5: nilai 55 -> REMIDI
Mahasiswa ke-6: nilai 78 -> LULUS
Mahasiswa ke-7: nilai 30 -> REMIDI
Mahasiswa ke-8: nilai 88 -> LULUS
Mahasiswa ke-9: nilai 60 -> LULUS
Mahasiswa ke-10: nilai 45 -> REMIDI


### <font color="red">Kegiatan 3</font>

In [ ]:
# int luas(int p, int l) { return p * l; }  <- versi C
def luas(p, l):
    return p * l

print(luas(4, 6))          # 24
print(luas(p=10, l=3))     # 30 -- argumen boleh disebut namanya

24
30


### <font color="red">Adaptasi kegiatan 3</font>

In [ ]:
# Adaptasi: fungsi untuk menghitung nilai akhir mahasiswa
def nilai_akhir(tugas, uts, uas):
    return tugas * 0.3 + uts * 0.3 + uas * 0.4

print(nilai_akhir(80, 75, 90)) # dipanggil dengan posisi
print(nilai_akhir(tugas=90, uts=85, uas=80)) # dipanggil dengan keyword argument

82.5
84.5


### <font color="red">Kegiatan 4</font>

In [ ]:
nilai = [78, 90, 66, 85]        # list -- array fleksibel
nilai.append(72)                # tambah data baru di belakang
print("Jumlah data :", len(nilai))
print("Rata-rata   :", sum(nilai) / len(nilai))
print("Dua teratas :", sorted(nilai)[-2:])    # slicing

mhs = {"nama": "Budi", "nilai": 85}     # dictionary
print(mhs["nama"], "mendapat", mhs["nilai"])

kelas = [
    {"nama": "Budi", "nilai": 85},
    {"nama": "Sari", "nilai": 92},
    {"nama": "Andi", "nilai": 58},
]
for m in kelas:
    status = "LULUS" if m["nilai"] >= 60 else "REMIDI"
    print(f"{m['nama']}: {m['nilai']} -> {status}")

Jumlah data : 5
Rata-rata   : 78.2
Dua teratas : [85, 90]
Budi mendapat 85
Budi: 85 -> LULUS
Sari: 92 -> LULUS
Andi: 58 -> REMIDI


### <font color="red">Adaptasi kegiatan 4</font>

In [ ]:
# Adaptasi: list dan dictionary dengan data nilai UAS + pengecekan predikat
nilai_uas = [70, 88, 95, 60, 45]
nilai_uas.append(100)                     # tambah satu data lagi
print("Jumlah data UAS:", len(nilai_uas))
print("Rata-rata UAS  :", sum(nilai_uas) / len(nilai_uas))
print("Tiga teratas   :", sorted(nilai_uas)[-3:])

kelas_uas = [
    {"nama": "Farid", "nilai": 85},
    {"nama": "Fadil", "nilai": 92},
    {"nama": "Arkan", "nilai": 58},
]
for m in kelas_uas:
    if m["nilai"] >= 85:
        predikat = "ISTIMEWA"
    elif m["nilai"] >= 60:
        predikat = "LULUS"
    else:
        predikat = "REMIDI"
    print(f"{m['nama']}: {m['nilai']} -> {predikat}")

Jumlah data UAS: 6
Rata-rata UAS  : 76.33333333333333
Tiga teratas   : [88, 95, 100]
Farid: 85 -> ISTIMEWA
Fadil: 92 -> ISTIMEWA
Arkan: 58 -> REMIDI


### <font color="red">Latihan 1</font>

In [ ]:
# Data nilai berdasarkan contoh program C di modul
nilai = [78, 90, 66, 85]
rata_rata = sum(nilai) / len(nilai)
print(f"Rata-rata:", rata_rata)

Rata-rata: 79.75


###<font color="red">penjelasan singkat latihan 1
**Konversi Rata-rata:** Baris sum(nilai) / len(nilai) Fungsi sum() menjumlahkan seluruh elemen list, sedangkan len() menghitung jumlah datanya.

**Contoh sum**()<font color="green"> (penjumlahan)<font color="black">: jika mengetik menulis sum() <font color="red">[78, 90, 66, 85]<font color="black">,Python akan langsung menghitung totalnya yaitu 319.

**Contoh len**() <font color="green">(length/panjang)<font color="Black"> :Dari list di atas, jumlah datanya ada 4 buah angka, maka len() akan menghasilkan angka 4.

> Add blockquote



### <font color="red">Latihan 2</font>

In [ ]:
def konversi_nilai(angka):
    if angka > 82:
        return "A"
    elif angka > 75:
        return "AB"
    elif angka > 67:
        return "B"
    elif angka > 59:
        return "BC"
    elif angka > 49:
        return "C"
    elif angka > 39:
        return "D"
    else:
        return "E"

print("1. Nilai 82 = ",konversi_nilai(82))
print("2. Nilai 75 = ",konversi_nilai(75))
print("3. Nilai 67 = ",konversi_nilai(67))


1. Nilai 82 =  AB
2. Nilai 75 =  B
3. Nilai 67 =  BC


###<font color="red">penjelasan singkat latihan 2
**(def)** : membungkus kode ke dalam sebuah Fungsi, jika memakai analogi (def) itu seperti sebuah mesin blender yang bisa mengubah buah(input) menjadi jus(output/return)  
**(if / elif / else)** : kontrol untuk meng eksekusi kode tertentu berdasarkan kondisi berdasarkan pemenuhan kriteria atau suatu kondisi logis,Jika suatu kondisi bernilai benar (True), blok kode di dalamnya dijalankan dan sisa kondisi di bawahnya diabaikan.

**if** angka > 82: mengecek apakah nilai lebih besar dari 82.

**elif** angka > 75: mengecek kondisi berikutnya jika rentang pertama tidak terpenuhi.

**else**: berfungsi sebagai penampung untuk kondisi terakhir (nilai di bawah ambang batas sebelumnya).

### <font color="red">Latihan 3</font>

In [ ]:
#list of dictionary berisi 5 mahasiswa
data_mahasiswa = [
    {"nama": "Farid", "nilai": 85},
    {"nama": "Fadil", "nilai": 92},
    {"nama": "Arkan", "nilai": 58},
    {"nama": "Ferdi", "nilai": 78},
    {"nama": "Adrian", "nilai": 45}
]
# 2. Menghitung rata-rata kelas
list_nilai = []
for mhs in data_mahasiswa:
    list_nilai.append(mhs["nilai"])

rata_kelas = sum(list_nilai) / len(list_nilai)
print(f"Rata-rata kelas: {rata_kelas}\n")

# 3. Mencetak nama di atas rata-rata dan mengecek huruf mutunya
print("Daftar mahasiswa dengan nilai di atas rata-rata:")
for mhs in data_mahasiswa:
    if mhs["nilai"] > rata_kelas:

        huruf = konversi_nilai(mhs["nilai"])
        print(f"- {mhs['nama']} (Nilai: {mhs['nilai']}, Mutu: {huruf})")

Rata-rata kelas: 71.6

Daftar mahasiswa dengan nilai di atas rata-rata:
- Farid (Nilai: 85, Mutu: A)
- Fadil (Nilai: 92, Mutu: A)
- Ferdi (Nilai: 78, Mutu: AB)


###<font color="red">penjelasan singkat latihan 3
**Membuat List of Dictionary (data_mahasiswa)**:
Variabel ini menampung data 5 mahasiswa dalam bentuk tabel data yang memuat kunci nama dan nilai  

**Menghitung Rata-rata Kelas**: Perulangan (for) digunakan untuk mengambil nilai setiap mahasiswa dan memasukkannya ke dalam list penampung menggunakan fungsi .append(). Rata-rata dihitung secara otomatis dengan membagi total penjumlahan nilai (sum) dengan jumlah data (len), lalu hasilnya (rata_kelas = 71.6).

**Mencetak mahasiswa di atas rata-rata beserta huruf mutu**: Loop kedua ini mengecek satu per satu: apakah mhs["nilai"] lebih besar dari rata_kelas?
jika iya, konversi_nilai(mhs["nilai"]) dipanggil untuk mengubah angka menjadi huruf mutu PNM (fungsi ini didefinisikan di Latihan 2, dengan aturan: A >82, AB >75, B >67, BC >59, C >49, D >39, E sisanya).
f"- {mhs['nama']} (Nilai: {mhs['nilai']}, Mutu: {huruf})" mencetak hasil dengan format rapi memakai f-string.